File: scripts/ga4.js
#!/usr/bin/env node
/**
 * Google Analytics 4 Data API Client — 零依赖 Node.js 版本
 *
 * 使用 Service Account（服务账号）JWT 认证，无需浏览器 OAuth。
 *
 * 前置步骤:
 *   1. Google Cloud Console → IAM → Service Accounts → 创建
 *   2. 下载 JSON 密钥，保存为 scripts/config/ga4_service_account.json
 *   3. GA4 后台 → 管理 → 用户管理 → 添加 Service Account 邮箱（至少"查看者"权限）
 *
 * 使用方法:
 *   node scripts/ga4.js --property 123456789 --days 28
 *
 * 注意:
 *   GA4 Property ID 是纯数字，在 GA4 设置 → 媒体资源详情中查看。
 */

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const TOKEN_URL = "https://oauth2.googleapis.com/token";
const GA4_API = "https://analyticsdata.googleapis.com/v1beta";
const SA_FILE = path.join(__dirname, "config", "ga4_service_account.json");

// ===================== JWT + Service Account =====================

function base64url(source) {
  return source.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

async function getAccessToken() {
  if (!fs.existsSync(SA_FILE)) {
    throw new Error(`Service Account 文件未找到: ${SA_FILE}\n请到 Google Cloud Console 创建并下载。`);
  }

  const sa = JSON.parse(fs.readFileSync(SA_FILE, "utf8"));
  const now = Math.floor(Date.now() / 1000);

  const header = JSON.stringify({ alg: "RS256", typ: "JWT" });
  const claims = JSON.stringify({
    iss: sa.client_email,
    scope: "https://www.googleapis.com/auth/analytics.readonly",
    aud: TOKEN_URL,
    iat: now,
    exp: now + 3600,
  });

  const signingInput = base64url(Buffer.from(header)) + "." + base64url(Buffer.from(claims));

  const sign = crypto.createSign("RSA-SHA256");
  sign.update(signingInput);
  const signature = sign.sign(sa.private_key, "base64");

  const jwt = signingInput + "." + base64url(Buffer.from(signature, "base64"));

  const res = await fetch(TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
      assertion: jwt,
    }),
  });

  const data = await res.json();
  if (!res.ok) throw new Error(`Token 获取失败: ${data.error_description || data.error}`);
  return data.access_token;
}

// ===================== API 调用 =====================

async function runReport(token, propertyId, days = 28) {
  const end = new Date().toISOString().split("T")[0];
  const start = new Date(Date.now() - days * 86400000).toISOString().split("T")[0];

  const body = {
    dateRanges: [{ startDate: start, endDate: end }],
    dimensions: [{ name: "pageTitle" }, { name: "pagePath" }],
    metrics: [
      { name: "sessions" },
      { name: "engagementRate" },
      { name: "averageEngagementTimePerSession" },
      { name: "bounceRate" },
      { name: "newUsers" },
      { name: "screenPageViews" },
    ],
    limit: "100",
  };

  const res = await fetch(`${GA4_API}/properties/${propertyId}:runReport`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  const data = await res.json();
  if (!res.ok) throw new Error(`GA4 API ${res.status}: ${data.error?.message || JSON.stringify(data)}`);

  const dimensions = data.dimensionHeaders?.map(h => h.name) || [];
  const metrics = data.metricHeaders?.map(h => h.name) || [];

  const pages = (data.rows || []).map(row => {
    const obj = {};
    row.dimensionValues?.forEach((v, i) => { obj[dimensions[i]] = v.value; });
    row.metricValues?.forEach((v, i) => {
      const val = parseFloat(v.value);
      obj[metrics[i]] = metrics[i].includes("Rate") ? Math.round(val * 10000) / 100 : val;
    });
    return obj;
  });

  return {
    propertyId,
    period: `${start} to ${end}`,
    totalPages: pages.length,
    pages,
  };
}

async function getTrafficSources(token, propertyId, days = 28) {
  const end = new Date().toISOString().split("T")[0];
  const start = new Date(Date.now() - days * 86400000).toISOString().split("T")[0];

  const body = {
    dateRanges: [{ startDate: start, endDate: end }],
    dimensions: [{ name: "sessionDefaultChannelGroup" }],
    metrics: [{ name: "sessions" }, { name: "engagementRate" }],
  };

  const res = await fetch(`${GA4_API}/properties/${propertyId}:runReport`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  const data = await res.json();
  if (!res.ok) throw new Error(`GA4 API ${res.status}: ${data.error?.message || JSON.stringify(data)}`);

  const sources = (data.rows || []).map(row => ({
    channel: row.dimensionValues[0]?.value,
    sessions: parseInt(row.metricValues[0]?.value, 10),
    engagementRate: Math.round(parseFloat(row.metricValues[1]?.value) * 10000) / 100,
  }));

  return { period: `${start} to ${end}`, sources };
}

// ===================== CLI =====================

function parseArgs(argv) {
  const args = { property: "", days: 28, traffic: false, all: false };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--property") args.property = argv[++i];
    else if (a === "--days") args.days = parseInt(argv[++i], 10);
    else if (a === "--traffic") args.traffic = true;
    else if (a === "--all") args.all = true;
  }
  return args;
}

(async () => {
  const args = parseArgs(process.argv);

  if (!args.property) {
    console.error("Usage: node scripts/ga4.js --property 123456789 [--days 28] [--traffic] [--all]");
    console.error("\n注意: 需要先将 Service Account JSON 保存到 scripts/config/ga4_service_account.json");
    console.error("      并把 Service Account 邮箱加入 GA4 的用户权限。");
    process.exit(1);
  }

  try {
    const token = await getAccessToken();
    const result = { propertyId: args.property };

    if (args.all || !args.traffic) {
      result.pageMetrics = await runReport(token, args.property, args.days);
    }
    if (args.all || args.traffic) {
      result.trafficSources = await getTrafficSources(token, args.property, args.days);
    }

    console.log(JSON.stringify(result, null, 2));
  } catch (e) {
    console.error("Error:", e.message);
    process.exit(1);
  }
})();

File: scripts/ga4.py
#!/usr/bin/env python3
"""
Google Analytics 4 Data API Client

免费，需要 OAuth 2.0 授权。
获取：页面互动率、平均互动时长、用户行为路径、流量来源。

使用方法:
    # 1. 到 Google Cloud Console 启用 GA4 Data API，下载 OAuth 凭证
    # 2. 将凭证保存为 scripts/config/ga4_credentials.json
    # 3. 运行:

    python scripts/ga4.py --property 123456789 --days 28

所需包:
    pip install google-analytics-data google-auth google-auth-oauthlib

注意:
    GA4 Property ID 是纯数字（如 123456789），在 GA4 设置中查看。
"""

import os
import sys
import json
import argparse
from datetime import datetime, timedelta

try:
    from google.analytics.data_v1beta import BetaAnalyticsDataClient
    from google.analytics.data_v1beta.types import (
        RunReportRequest,
        DateRange,
        Dimension,
        Metric,
    )
    from google.oauth2.credentials import Credentials
    from google_auth_oauthlib.flow import InstalledAppFlow
except ImportError:
    print("请先安装依赖: pip install google-analytics-data google-auth google-auth-oauthlib")
    sys.exit(1)

SCOPES = ["https://www.googleapis.com/auth/analytics.readonly"]
CREDENTIALS_FILE = os.path.join(os.path.dirname(__file__), "config", "ga4_credentials.json")
TOKEN_FILE = os.path.join(os.path.dirname(__file__), "config", "ga4_token.json")


def get_client():
    """获取已授权的 GA4 Data API client。"""
    creds = None

    if os.path.exists(TOKEN_FILE):
        creds = Credentials.from_authorized_user_file(TOKEN_FILE, SCOPES)

    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            if not os.path.exists(CREDENTIALS_FILE):
                print(f"错误: 未找到凭证文件 {CREDENTIALS_FILE}")
                print("请到 Google Cloud Console 下载 OAuth 凭证并重命名为 ga4_credentials.json")
                sys.exit(1)
            flow = InstalledAppFlow.from_client_secrets_file(CREDENTIALS_FILE, SCOPES)
            creds = flow.run_local_server(port=0)

        os.makedirs(os.path.dirname(TOKEN_FILE), exist_ok=True)
        with open(TOKEN_FILE, "w") as f:
            f.write(creds.to_json())

    return BetaAnalyticsDataClient(credentials=creds)


def run_report(client, property_id: str, days: int = 28):
    """获取核心行为指标。"""
    end_date = datetime.now().strftime("%Y-%m-%d")
    start_date = (datetime.now() - timedelta(days=days)).strftime("%Y-%m-%d")

    request = RunReportRequest(
        property=f"properties/{property_id}",
        date_ranges=[DateRange(start_date=start_date, end_date=end_date)],
        dimensions=[
            Dimension(name="pageTitle"),
            Dimension(name="pagePath"),
        ],
        metrics=[
            Metric(name="sessions"),
            Metric(name="engagementRate"),
            Metric(name="averageEngagementTimePerSession"),
            Metric(name="bounceRate"),
            Metric(name="newUsers"),
            Metric(name="screenPageViews"),
        ],
        limit=100,
    )

    response = client.run_report(request)

    pages = []
    for row in response.rows:
        pages.append({
            "pageTitle": row.dimension_values[0].value,
            "pagePath": row.dimension_values[1].value,
            "sessions": int(row.metric_values[0].value),
            "engagementRate": round(float(row.metric_values[1].value) * 100, 2),
            "avgEngagementTimeSec": round(float(row.metric_values[2].value), 1),
            "bounceRate": round(float(row.metric_values[3].value) * 100, 2),
            "newUsers": int(row.metric_values[4].value),
            "pageViews": int(row.metric_values[5].value),
        })

    return {
        "propertyId": property_id,
        "period": f"{start_date} to {end_date}",
        "totalPages": len(pages),
        "pages": pages,
    }


def get_traffic_sources(client, property_id: str, days: int = 28):
    """获取流量来源分布。"""
    end_date = datetime.now().strftime("%Y-%m-%d")
    start_date = (datetime.now() - timedelta(days=days)).strftime("%Y-%m-%d")

    request = RunReportRequest(
        property=f"properties/{property_id}",
        date_ranges=[DateRange(start_date=start_date, end_date=end_date)],
        dimensions=[Dimension(name="sessionDefaultChannelGroup")],
        metrics=[
            Metric(name="sessions"),
            Metric(name="engagementRate"),
        ],
    )

    response = client.run_report(request)

    sources = []
    for row in response.rows:
        sources.append({
            "channel": row.dimension_values[0].value,
            "sessions": int(row.metric_values[0].value),
            "engagementRate": round(float(row.metric_values[1].value) * 100, 2),
        })

    return {
        "period": f"{start_date} to {end_date}",
        "sources": sources,
    }


def main():
    parser = argparse.ArgumentParser(description="Google Analytics 4 Data API Client")
    parser.add_argument("--property", required=True, help="GA4 Property ID（纯数字）")
    parser.add_argument("--days", type=int, default=28, help="查询天数（默认 28）")
    parser.add_argument("--traffic", action="store_true", help="获取流量来源")
    parser.add_argument("--all", action="store_true", help="获取所有数据")

    args = parser.parse_args()

    try:
        client = get_client()
        result = {"propertyId": args.property}

        if args.all or not args.traffic:
            result["pageMetrics"] = run_report(client, args.property, args.days)

        if args.all or args.traffic:
            result["trafficSources"] = get_traffic_sources(client, args.property, args.days)

        print(json.dumps(result, indent=2, ensure_ascii=False))

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
